mycodo/controllers/controller_pid.py

```diff
@@ -252,18 +250,56 @@ def check_pid(self):
-                    new_setpoint, ended = calculate_method_setpoint(
-                        self.setpoint_tracking_id,
-                        PID,
-                        this_pid,
-                        Method,
-                        MethodData,
-                        self.logger)
-                    if ended:
-                        self.method_start_act = 'Ended'
+
+                    now = datetime.datetime.now()
+
+                    if this_pid.method_start_time is None or this_pid.method_start_time == 'Ready':
+                        this_pid.method_start_time = now
+
+                    method = load_method(self.setpoint_tracking_id, self.logger)
+                    new_setpoint, finished = method.calculate_setpoint(now, this_pid.method_start_time)
+                    self.logger.debug("Method {} {} {} {}".format(self.setpoint_tracking_id, method, now, this_pid.method_start_time))
+
+                    # Check if method_end_time is not None
+                    end_time_reached = False
+                    if this_pid.method_end_time:
+                        # Convert time string to datetime object
+                        end_time = datetime.datetime.strptime(
+                            str(this_pid.method_end_time), '%Y-%m-%d %H:%M:%S.%f')
+                        if now > end_time:
+                            end_time_reached = True
+
+                    if finished or end_time_reached:
+                        # point in time is out of method range
+                        with session_scope(MYCODO_DB_PATH) as db_session:
+                            # Overwrite this_controller with committable connection
+                            this_pid = db_session.query(PID).filter(
+                                PID.unique_id == self.unique_id).first()
+
+                            if not end_time_reached and method.should_restart(now, this_pid.method_start_time):
+                                self.logger.debug("Restart")
+                                # Method has been instructed to restart
+                                this_pid.method_start_time = now
+                                # Calculate restart value
+                                new_setpoint, finished = method.calculate_setpoint(now, this_pid.method_start_time)
+                            else:
+                                self.logger.debug("Ended")
+                                # Duration method has ended, reset method_start_time locally and in DB
+                                this_pid.method_start_time = 'Ended'
+                                this_pid.method_end_time = None
+                                this_pid.is_activated = False
+                                db_session.commit()
+
+                                self.is_activated = False
+                                self.stop_controller()
+
+                            db_session.commit()
```

Comment:
Previous implementation did database access for this partly in the controller and partly in the method. the part that was done in method has been migrated to the controllers here.